### ISSUE TITLE
Thundering herd at expiration time

### ISSUE THREAD

**Issue by sysrqb** [chris-wood/draft-group-privacypass-k-check#8]
Consider adding a deployment consideration comment about how to avoid the thundering herd problem.

**Comment by bemasc** [comment]
The best solution I've thought of is to define each profile as providing a "ladder" comprising a pair of resources (perhaps distinguished by a request header).  The client fetches both, which are cached independently, and verifies that that they indicate the same configurations during the overlap in their cache lifetimes.  After the first expiration, the client picks a random time before the second one expires to refresh the first one.  This ensures that refresh requests are not clustered near expiration times, and allows us to continue using the Mirror Protocol as-is.  (Mirrors do not need to be aware of the ladder semantics.)  However, it doubles the number of requests, and complicates the deployment for origins.

**Comment by tfpauly** [comment]
Another approach is to generally recommend that clients fetch resources from the mirror as they need to use the resource (if I don't need to fetch a token, then I don't check the key when the key expires). For these less frequent cases, they shouldn't have a thundering herd.

For resources that are used very frequently, clients should avoid fetching the mirror resource at the exact time of expiration, and should have some jitter around that (potentially aligned to when they'd fetch a resource based on use anyway)

**Comment by bemasc** [comment]
> For these less frequent cases, they shouldn't have a thundering herd.

Yes: if the use frequency is long relative to the cache lifetime, then we don't expect clustering of refreshes at expiration points.  (Assuming on-demand validation, which seems like the reasonable default despite some concerns about timing correlations.)

> For resources that are used very frequently, clients should avoid fetching the mirror resource at the exact time of expiration, and should have some jitter around that

I don't think this works.  Fetches prior to the expiration time don't help: they just return a mirror resource that is about to expire.  Using the fetched resource after the expiration time is unsafe in two ways: it may be invalid (leading to failure) and you may be the only client still using it (leading to deanonymization).

**Comment by sysrqb** [comment]
Thanks for raising this issue Ben, and giving some guidance on how to solve it. I wrote some text around this. It takes a simpler approach that puts some responsibility on the service/origin, instead of describing a more complicated client-side chaining/ladder solution. Some implementations may benefit from chaining multiple resources, with different expiration periods, but that seems out of scope for the core protocol at this time. If there is interest in specifying this from the group, then we will discuss adding it in a later version of the draft.

**Comment by sysrqb** [comment]
Happy to continue the discussion from https://github.com/chris-wood/draft-group-privacypass-consistency-mirror/pull/21#discussion_r1365720273 if there is more that we need to consider.

**Comment by bemasc** [comment]
Ultimately, the options I see are:

* There is only one copy of the resource in use.  Thundering herd at replacement time.
* There are two or more copies of the resource in use with staggered lifetimes.  No thundering herd, but someone has to check that both copies are equivalent during the overlap in their lifetimes.
  - Clients fetch both and check: "ladder" strategy as above (https://github.com/chris-wood/draft-group-privacypass-consistency-mirror/issues/8#issuecomment-1711918025)
  - The mirror understands the resource and performs the equivalence check before updating its cache (a "smart mirror" strategy).

The "ladder" strategy seems preferable to me because the "smart mirror" approach has some issues:

* It creates ossification risk (mirrors need to add support for new formats before they can be used).
* It doesn't support novel uses of mirrors beyond consistency checking.
* It creates a risk of consistency failures if the mirror and the clients have different interpretations of the resource.
